From 4a1e3440f4d208fe2c36c80ddcdcfa7cb15b7f35 Mon Sep 17 00:00:00 2001
From: OpenEmbedded <mario.schuknecht@dresearch-fe.de>
Date: Mon, 20 Feb 2023 16:21:41 +0100
Subject: [PATCH] Add parameter to module 'brcmfmac' HYP-29550

The Laird module with chip id 4373 must be started with the correct
regulatory domain. This value can now be passed as a module parameter.

Signed-off-by: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
---
 .../broadcom/brcm80211/brcmfmac/cfg80211.c        | 15 +++++++++++++++
 .../wireless/broadcom/brcm80211/brcmfmac/common.c |  9 +++++++++
 .../wireless/broadcom/brcm80211/brcmfmac/common.h |  2 ++
 3 files changed, 26 insertions(+)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
index 9db12ffd2ff8..029cb715c1d9 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
@@ -7609,6 +7609,21 @@ struct brcmf_cfg80211_info *brcmf_cfg80211_attach(struct brcmf_pub *drvr,
 	ndev->ieee80211_ptr = &vif->wdev;
 	SET_NETDEV_DEV(ndev, wiphy_dev(cfg->wiphy));
 
+	if (strlen(drvr->settings->regdomain) != 0) {
+		struct brcmf_fil_country_le ccreq;
+		memset(&ccreq, 0, sizeof(ccreq));
+		ccreq.rev = -1;
+
+		memcpy(ccreq.country_abbrev, drvr->settings->regdomain, BRCMF_COUNTRY_BUF_SZ);
+
+		err = brcmf_fil_iovar_data_set(ifp, "country", &ccreq, sizeof(ccreq));
+		if (err) {
+			bphy_err(drvr, "Regulatory domain %s not supported, aborting!\n", drvr->settings->regdomain);
+			goto wiphy_out;
+		}
+		brcmf_info("Using regulatory domain %s\n", drvr->settings->regdomain);
+	}
+
 	err = wl_init_priv(cfg);
 	if (err) {
 		bphy_err(drvr, "Failed to init iwm_priv (%d)\n", err);
diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
index e3758bd86acf..2437c3bce972 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
@@ -67,6 +67,11 @@ static int brcmf_iapp_enable;
 module_param_named(iapp, brcmf_iapp_enable, int, 0);
 MODULE_PARM_DESC(iapp, "Enable partial support for the obsoleted Inter-Access Point Protocol");
 
+static char brcmf_regdomain[BRCMF_REGDOMAIN_LEN] = "DE";
+module_param_string(regdomain, brcmf_regdomain,
+                    BRCMF_REGDOMAIN_LEN, 0400);
+MODULE_PARM_DESC(regdomain, "Regulatory domain/country code");
+
 #ifdef DEBUG
 /* always succeed brcmf_bus_started() */
 static int brcmf_ignore_probe_fail;
@@ -417,6 +422,10 @@ struct brcmf_mp_device *brcmf_get_module_param(struct device *dev,
 	settings->ignore_probe_fail = !!brcmf_ignore_probe_fail;
 #endif
 
+	// Laird - Copy regulory domain module parameter, subject to
+	// override by DT
+	strlcpy(settings->regdomain, brcmf_regdomain, BRCMF_REGDOMAIN_LEN);
+
 	if (bus_type == BRCMF_BUSTYPE_SDIO)
 		settings->bus.sdio.txglomsz = brcmf_sdiod_txglomsz;
 
diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h
index 8b5f49997c8b..63863d480bc8 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h
@@ -10,6 +10,7 @@
 #include "fwil_types.h"
 
 #define BRCMF_FW_ALTPATH_LEN			256
+#define BRCMF_REGDOMAIN_LEN			10
 
 /* Definitions for the module global and device specific settings are defined
  * here. Two structs are used for them. brcmf_mp_global_t and brcmf_mp_device.
@@ -49,6 +50,7 @@ struct brcmf_mp_device {
 	bool		iapp;
 	bool		ignore_probe_fail;
 	struct brcmfmac_pd_cc *country_codes;
+	char		regdomain[BRCMF_REGDOMAIN_LEN];
 	const char	*board_type;
 	union {
 		struct brcmfmac_sdio_pd sdio;
-- 
2.35.3

